### PR TITLE
Edge app workflow

### PR DESCRIPTION
**Pull request by sergey-borovkov**
## What does this PR do?
Add playlist commands and edge app commands.

Implements shortcuts that allow to append and prepend assets to playlists and edit any playlist as a JSON file.

Initial support for edge apps:
* Edge app creation (in the database and template on disk)
* Edge app upload command that scans for any changes, creates a new version of app if necessary, and uploads all content.
* It then publishes that version (makes it immutable) and instantiates it if needed. At that point edge app asset can be scheduled to the playlist.

What's missing:
* Setting settings
* Releasing stable/candidate version of edge apps.

Please refer to the cli.rs for details of every command - they are self documented there and the description will show up when running the app.
## GitHub issue or Phabricator ticket number?
T6977 and T6787
## How has this been tested?
* Every command was tested manually.
* Unit tests check for all workflows (basic usage without covering all edge cases as that would be too big of a task).
* 
## Checklist before merging

- [x] If have added tests.
- [x] Is this a new feature? If yes, please write one phrase about this update.
- [ ] I've attached relevant screenshots (if relevant).

### CODE REVIEW

File: data/index.html [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <title>Your Edge App name</title>
+    <link rel="stylesheet" href="style.css">
+    <script src="script.js"></script>
```
Comment (vpetersson) [inline data/index.html:7]: We should probably add something that users the metadata. Like

`Hello World, I'm $SCREEN_NAME`

Comment (sergey-borovkov) [reply data/index.html:7]: Sure, will change that a bit later

File: data/index.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <title>Your Edge App name</title>
```
[the remark this answers is not in the document]
Comment (sergey-borovkov) [reply data/index.html:5]: ok

**Review by renatgalimov — COMMENTED** [review]
@sergey-borovkov please fix all clippy errors.

File: src/commands/edge_app.rs [hunk 3]
```diff
@@ -1,227 +1,558 @@
...
+    let mut changes_detected = false;
+
+    let mut remote_iter = remote_files.iter().peekable();
+
+    for local_file in local_files {
+        let local_signature = &local_file.signature;
+
+        while let Some(remote_file) = remote_iter.peek() {
+            if &remote_file.signature == local_signature {
+                copies.push(remote_file.clone().clone());
```
[the remark this answers is not in the document]
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:494]: clone().clone()... yeah need to figure out how to simplify it but remote_file is && reference that's what causes the issue

File: src/commands/playlist.rs [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+        if playlist_items.len() != 1 {
+            return Err(CommandError::MissingField);
+        }
+        let position = playlist_items[0].position + POSITION_MULTIPLIER;
+        let payload = json!([{
+            "playlist_id": playlist_uuid,
+            "asset_id": asset_uuid,
+            "duration": duration,
+            "position": position
+        }]);
```
[the remark this answers is not in the document]
Comment (sergey-borovkov) [reply src/commands/playlist.rs:173]: no, position is not unique. if pos is equal it will sort by ulid.

File: src/commands/playlist.rs [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+
+        // then post request to create new playlist items
+        let post_mock = mock_server.mock(|when, then| {
+            when.method(POST)
+                .path("/v4/playlist_items")
+                .header("Authorization", "Token token")
+                .json_body(items_request);
+            then.status(201).json_body(json!({}));
+        });
+
```
[the remark this answers is not in the document]
Comment (sergey-borovkov) [reply src/commands/playlist.rs:584]: hmm, perhaps. but I don't want to deal with this now - that's how other tests work as well. Mb during gap week

**Review by renatgalimov — APPROVED** [review]

File: data/index.html [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <script src="screenly.js"></script>
+    <title>Hello world</title>
+  </head>
+  <body>
+    <p id="message">Hello world from <span id="hostname-placeholder">%1</span></p>
```
Comment (korvyashka) [inline data/index.html:9]: %1 is not needed here.

File: src/cli.rs [hunk 7]
```diff
@@ -100,6 +108,71 @@ pub enum ScreenCommands {
     },
 }
 
+#[derive(Subcommand, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PlaylistCommands {
```
Comment (korvyashka) [inline src/cli.rs:112]: playlist changes as well?

File: src/cli.rs [hunk 8]
```diff
@@ -527,22 +696,87 @@ pub fn handle_cli_asset_command(command: &AssetCommands) {
                 }
             }
         }
+        AssetCommands::BearerAuth { uuid, token } => {
```
Comment (korvyashka) [inline src/cli.rs:699]: New asset command?

File: src/commands/edge_app.rs [hunk 9]
```diff
@@ -1,227 +1,777 @@
...
         Self { authentication }
     }
 
-    pub fn init(self, path: &Path) -> Result<(), CommandError> {
-        let mut object = serde_yaml::to_value(EdgeAppManifest::default())?;
-        let map = object.as_mapping_mut().ok_or(CommandError::MissingField)?;
+    pub fn create(&self, name: &str, path: &Path) -> Result<(), CommandError> {
+        let response = commands::post(
+            &self.authentication,
+            "v4/edge-apps?select=id,name",
```
Comment (korvyashka) [inline src/commands/edge_app.rs:44]: Hm - name will be the one you sent certainly.

File: src/commands/edge_app.rs [hunk 10]
```diff
@@ -1,227 +1,777 @@
...
-        // following fields will be generated server side when publishing.
-        map.remove("id");
+        let yaml = serde_yaml::to_string(&manifest)?;
+        let manifest_file = File::create(path)?;
+        write!(&manifest_file, "{yaml}")?;
 
-        let yaml = serde_yaml::to_string(map)?;
-        let input = File::create(path)?;
-        write!(&input, "{yaml}")?;
+        let index_html_template = include_str!("../../data/index.html");
```
Comment (korvyashka) [inline src/commands/edge_app.rs:71]: Hm? Why it index.html - shouldn't it be arbitrary files from user?

File: src/commands/edge_app.rs [hunk 11]
```diff
@@ -1,227 +1,777 @@
...
 
-    pub fn publish(self, path: &Path) -> Result<EdgeAppManifest, CommandError> {
-        let url = format!(
-            "{}/v4/edge_apps?select=id,name,version,description,icon,author,homepage_url",
-            &self.authentication.config.url
-        );
+    pub fn list(&self) -> Result<EdgeApps, CommandError> {
+        Ok(EdgeApps::new(commands::get(
+            &self.authentication,
+            "v4/edge-apps",
```
Comment (korvyashka) [inline src/commands/edge_app.rs:87]: May be list selected fields

File: src/commands/edge_app.rs [hunk 12]
```diff
@@ -1,227 +1,777 @@
...
+
+        self.upload_changed_files(edge_app_dir, &manifest, &changed_files)?;
+        debug!("Files uploaded");
+
+        self.ensure_assets_processing_finished(&manifest)?;
+        // now we freeze it by publishing it
+        self.publish(&manifest)?;
+        debug!("Edge app published.");
+
+        let root_asset_id = match self.get_root_edge_app_asset(&manifest) {
```
Comment (korvyashka) [inline src/commands/edge_app.rs:166]: We have different naming here and on backend.
What you call root edge app asset - is an instance of an edge app in backend terminology.

File: src/commands/edge_app.rs [hunk 13]
```diff
@@ -1,227 +1,777 @@
...
+    fn ensure_assets_processing_finished(
+        &self,
+        manifest: &EdgeAppManifest,
+    ) -> Result<(), CommandError> {
+        const SLEEP_TIME: u64 = 2;
+        loop {
+            let value = commands::get(
+                &self.authentication,
+                &format!(
+                    "v4/assets?select=status&app_id=eq.{}&app_revision=eq.{}&status=neq.finished&limit=1",
```
Comment (korvyashka) [inline src/commands/edge_app.rs:275]: Could you add TODO comment here regarding error status.

File: src/commands/edge_app.rs [hunk 14]
```diff
@@ -1,227 +1,777 @@
...
+                ),
+            )?;
+            debug!("ensure_assets_processing_finished: {:?}", &value);
+
+            if let Some(array) = value.as_array() {
+                if array.is_empty() {
+                    break;
+                }
+            }
+            thread::sleep(Duration::from_secs(SLEEP_TIME));
```
Comment (korvyashka) [inline src/commands/edge_app.rs:286]: Let not sleep forever, may be fail after 20-30 secs?

**Review by korvyashka — APPROVED** [review]

File: src/commands/edge_app.rs (hunk 9 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:44]: yeah - and easier to parse it from response then

File: src/commands/edge_app.rs (hunk 10 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:71]: index.html is an entry point to our edge apps at the moment. So we give them a small template.

File: src/commands/edge_app.rs (hunk 12 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:166]: that's what you called it as well when we were on the call...

File: data/index.html (hunk 6 above)
Comment (sergey-borovkov) [reply data/index.html:9]: are you sure? I tested it on the device %1 was not displayed

File: src/cli.rs (hunk 7 above)
Comment (sergey-borovkov) [reply src/cli.rs:112]: yes, playlist commands were added in this PR because I had too many changes tied in here

File: src/cli.rs (hunk 8 above)
Comment (sergey-borovkov) [reply src/cli.rs:699]: hmm no, rebase issue I think... need to merge master to this branch

File: src/commands/edge_app.rs (hunk 9 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:44]: Fixed - merge issue

File: src/cli.rs (hunk 8 above)
Comment (sergey-borovkov) [reply src/cli.rs:699]: fixed

File: src/commands/edge_app.rs (hunk 13 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:285]: done

File: src/commands/edge_app.rs (hunk 14 above)
Comment (sergey-borovkov) [reply src/commands/edge_app.rs:296]: done
